[WIP] Prefer using the network interface with the default route on Windows

sebsoto opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR corrects a difference in behavior between Linux and Windows, when choosing a network interface.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #104269

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

sebsoto commented:
/sig windows

jsturtevant commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:
```diff
@@ -244,23 +244,22 @@ func getMatchingGlobalIP(addrs []net.Addr, family AddressFamily) (net.IP, error)
-	intf, err := nw.InterfaceByName(intfName)
-	if err != nil {
-		return nil, err
+func getIPFromInterface(intf *net.Interface, forFamily AddressFamily, nw networkInterfacer) (net.IP, error) {
```
I don't follow why this change was necessary.  It looks like the core change is the `chooseHostInterface` to be per OS.

sebsoto replied:
I made this change to reduce repeated code.
I needed to use the same logic to find a valid IP, but unlike the Linux routing table which has the name of the interface, the Windows routing table contains the index of the interface.

I've just pushed changes which have reduced the delta of the changes I made, but I am still touching that function.

jsturtevant commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go:
```diff
@@ -0,0 +1,86 @@
+	// https://docs.microsoft.com/en-us/powershell/module/nettcpip/get-netroute?view=windowsserver2019-ps#example-4--get-the-next-hop-for-the-default-route
+	// -DestinationPrefix values of "0.0.0.0/0" and "::0/0", return entries with the NextHop value being the default
+	// gateway for IPv4, and IPv6, respectively.
+	cmd := exec.Command("powershell", "/c", `Get-NetRoute -ErrorAction Ignore -DestinationPrefix ("0.0.0.0/0","::0/0") | Format-Table -Property "ifIndex","NextHop" -HideTableHeaders`)
```
it would be preferable to use golang to avoid creating a process by calling out to powershell if possible

sebsoto replied:
Agreed, changed to using the [getipforwardentry2](https://docs.microsoft.com/en-us/windows/win32/api/netioapi/nf-netioapi-getipforwardentry2) function.

jsturtevant commented:
/priority important-longterm

sebsoto committed:
Find default routes on Windows

This commit makes it so that on Windows ChooseHostInterface() behaves
more like it does on Linux. When ran on Windows the function will now
prefer an interface that is associated with the default route, found via
a routing table lookup.

aojea commented:
/assign

aojea commented:
/ok-to-test
/kind bug

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go:
```diff
@@ -0,0 +1,142 @@
+	for _, family := range searchFamilies {
+		route, err := getDefaultRouteEntry(family)
+		if err != nil {
+			return nil, err
```
do you want to exit or continue and try the other family?

sebsoto replied:
Changed it to continue

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/net/interface_windows.go:
```diff
@@ -0,0 +1,142 @@
+		}
+		intf, err := net.InterfaceByIndex(int(route.InterfaceIndex))
+		if err != nil {
+			return nil, err
```
ditto, continue or exit?

sebsoto replied:
Changed it to continue

sebsoto committed:
Generated: hack/update-vendor.sh

dims commented:
please reopen if you still need this WIP PR. Trying to declutter the open PR(s)
